### PR TITLE
Initial framework for collecting anonymous usage telemetry

### LINKED ISSUE #2212: Create OTel exporter for `oss_telemetry`

### LINKED ISSUE #2213: Send `ping` telemetry to server

### LINKED ISSUE #2214: Create function/macro for recording telemetry

### LINKED ISSUE #2215: Provide a way to disable telemetry via CLI or `spicepod.yaml`

### LINKED ISSUE #2220: Implement `query_count` telemetry

### PR DESCRIPTION
**Pull request by phillipleblanc**
## 🗣 Description

Creates the initial framework for sending anonymous usage telemetry to the Spice AI servers. This telemetry is enabled by default, but can be disabled in one of 3 ways:

1. Building the binary without the `anonymous_telemetry` feature flag, which will remove the code that exports the telemetry completely.
2. Passing the `--telemetry-enabled false` flag as a CLI parameter.
3. Setting the following runtime config in `spicepod.yaml`:

```yaml
runtime:
  telemetry:
    enabled: false
```

Telemetry is sent to the server once an hour. The telemetry is sent in Arrow format via the Apache Arrow Flight `DoPut` RPC call.

This PR also tweaks the `FlightClient` to allow an anonymous mode.

## 🔨 Related Issues

Closes #2202
Closes #2213
Closes #2214 
Closes #2215
Closes #2220
Part of #1715

### CODE REVIEW

File: crates/telemetry/src/exporter.rs [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+#[async_trait]
+impl otel_arrow::ArrowExporter for AnonymousTelemetryExporter {
+    async fn export(&self, metrics: RecordBatch) -> Result<(), MetricsError> {
+        let Some(mut flight_client) = self.flight_client.clone() else {
+            return Ok(());
+        };
+
+        if let Err(e) = flight_client.publish("oss_telemetry", vec![metrics]).await {
+            tracing::error!("Unable to publish anonymous telemetry: {e}");
```
Comment (sgrebnov) [inline crates/telemetry/src/exporter.rs:70]: Not sure it is the best practice to report this as an error. Understand it is very convenient for anonymous telemetry feature debugging/troubleshooting but it seems other popular tools handle situation gracefully without showing errors to the user

**Review by sgrebnov — DISMISSED** [review]
~`telemetry_enabled` and `runtime.telemetry` names might confuse and refer to general metrics (not anonymous_telemetry). Might be better to be more explicit and name it `anonymous_telemetry_enabled` and `runtime.anonymous_telemetry`?~

Reviewed naming in popular tools and it is just `telemetry`👍

File: crates/telemetry/src/exporter.rs (hunk 1 above)
Comment (phillipleblanc) [reply crates/telemetry/src/exporter.rs:70]: This won't show by default, the `telemetry` crate isn't enabled in the default `SPICED_LOG`.

**Review by peasee — DISMISSED** [review]

File: crates/telemetry/src/anonymous.rs [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    let provider = SdkMeterProvider::builder()
+        .with_resource(resource)
+        .with_reader(periodic_reader)
+        .build();
+
+    let _ = crate::meter::METER_PROVIDER.get_or_init(|| GlobalMeterProvider::new(provider));
+
+    crate::meter::METER.u64_counter("start").init().add(1, &[]);
+
+    tracing::debug!("Started anonymous telemetry collection to {}", *ENDPOINT);
```
Comment (y-f-u) [inline crates/telemetry/src/anonymous.rs:63]: Does this tracing level needs to be more visible? As to let user know this is happening?

**Review by y-f-u — APPROVED** [review]

File: crates/telemetry/src/anonymous.rs (hunk 2 above)
Comment (phillipleblanc) [reply crates/telemetry/src/anonymous.rs:63]: No, this was just for me to make sure it was going to the right place.

File: crates/telemetry/src/anonymous.rs [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        .build();
+
+    let provider = SdkMeterProvider::builder()
+        .with_resource(resource)
+        .with_reader(periodic_reader)
+        .build();
+
+    let _ = crate::meter::METER_PROVIDER.get_or_init(|| GlobalMeterProvider::new(provider));
+
+    crate::meter::METER.u64_counter("start").init().add(1, &[]);
```
Comment (Jeadie) [inline crates/telemetry/src/anonymous.rs:61]: `"start"` signalling the telemetry has started? Could be better name.

Comment (phillipleblanc) [reply crates/telemetry/src/anonymous.rs:61]: Yeah, I'm probably going to remove this once I add the other metrics in.

File: crates/telemetry/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+use opentelemetry::metrics::Counter;
+use std::sync::LazyLock;
+
+#[cfg(feature = "anonymous_telemetry")]
+pub mod anonymous;
+mod exporter;
+mod meter;
+
+pub static QUERY_COUNT: LazyLock<Counter<u64>> = LazyLock::new(|| {
+    METER
```
Comment (Jeadie) [inline crates/telemetry/src/lib.rs:27]: OTel doesn't make us do it, this is us standardising what/how we meter, right?

Comment (phillipleblanc) [reply crates/telemetry/src/lib.rs:27]: It is actually required to use OTel to define the metrics up front.

**Pull request merged**